### DISCUSSION TITLE
AtomicServiceInterface doesn't rollback on exception?

### DISCUSSION THREAD

**Discussion by JavierFerrerGomez** [bavix/laravel-wallet#976] [General]
I didn't open this as an issue because I am not sure I understood how AtomicServiceInterface should work. If this is indeed a bug I will open an issue with a proper working test case.

When using **AtomicServiceInterface** I am expecting that in the case of an exception all operations already applied to the DB will be rollbacked, but that is not the case. This is the relevant code in my user model:

```php
class User extends Authenticatable implements MustVerifyEmail
{
    use HasApiTokens;
    use HasFactory;
    use Notifiable;
    use HasWallets;

    public function depositWalletTransaction(
        int $amount,
        string $slug,
        callable $onSuccessFn = null
    ) {

        $wallet = $this->getWallet($slug);
        app(AtomicServiceInterface::class)->block($wallet, function () use ($wallet, $amount, $onSuccessFn) {
            $walletTransaction = $wallet->deposit($amount);
            if ($onSuccessFn) {
                $onSuccessFn($walletTransaction);
            }
        });
    }
}
```
Then in my test case I have:
```php
    #[Test]
    public function transaction_deposit_should_rollback(): void
    {
        $user = $this->getTestUser(); // balance here is 10000

        try {            
            $user->depositWalletTransaction(10, "global", function ($transaction) use ($user) {
                $user->email = "modified@iamtest.com";
                $user->save();
                throw new GameWalletException('ERROR!');
            });
        } catch(\Exception $e) {
            $this->assertInstanceOf(GameWalletException::class, $e);
            $user = User::find(1)->first();
            assertEquals(10000, $user->getConfirmedGlobalBalanceInt());
            assertEquals('testy@original.com', $user->email);
        }
    }
```

When I runt the test, in my catch statement the current balance is **10010** (_getConfirmedGlobalBalanceInt_ calls refreshBalance internally) and the email address is **modified@iamtest.com**. What am I missing?

**Comment by rez1dent3** [comment]
Hello. I checked it. Everything works as expected. Here is the PR: https://github.com/bavix/laravel-wallet/pull/977/files

I am 99.99% sure that a transaction is created before your test and atomic does not create a new one. I suggest you add a check that the transaction level is zero.

Example:
```diff
    #[Test]
    public function transaction_deposit_should_rollback(): void
    {
+        self::assertSame(0, \Illuminate\Support\Facades\DB::transactionLevel());
```

I think the test will start to fail on this line.
